### PR TITLE
Prohibit casts between fat pointers to different traits

### LINKED ISSUE #26391: Clarify cast rules (tracking issue for RFC 1052)
Much of the implementation work for rust-lang/rfcs#1052 has been done, but there remain some corner cases to fix:

- [ ] Casts between raw pointers to different traits are accepted but should not be (e.g. http://is.gd/1jCOSz)
- [ ] Double-check other cases, ensure adequate tests exist

### PR DESCRIPTION
**Pull request by arielb1**
This makes them compliant with the new version of RFC 401 (i.e.
    RFC 1052).

Fixes #26391. I _hope_ the tests I have are enough.

This is a [breaking-change]

r? @nrc

### CODE REVIEW

**Comment by nrc** [comment]
ping @rust-lang/lang because of the breaking change

**Comment by nrc** [comment]
Also cc @brson to see if we can test this on crater

File: src/test/run-pass/cast-rfc0401.rs [hunk 1]
```diff
@@ -99,6 +99,12 @@ fn main()
     let l_via_str = unsafe{&*(s as *const [u8])};
     assert_eq!(&l, l_via_str);
 
+    // ptr-ptr-cast (both vk=Length, check length is preserved)
```
Comment (nrc) [inline src/test/run-pass/cast-rfc0401.rs]: nit: spell out vk in this comment

**Comment by nrc** [comment]
r+ for the code (modulo nit). We should discuss the breaking change aspect before landing this though

**Comment by arielb1** [comment]
@nrc 

addressed. I think the breaking change was part of the RFC (this is a 1.2 -> 1.2 breaking change).

**Comment by brson** [comment]
Crater is started on this.

**Comment by nrc** [comment]
Since this is only a breaking change within a version, I think it is OK to land. If Crater finds anything horrible some how, we can back out.

@bors: r+

**Comment by brson** [comment]
[crater says no regressions](https://gist.github.com/brson/32c6ee090a0d9c8f161b)
